forgelib/events.go

```diff
@@ -19,23 +19,45 @@ func (s *Stack) ListEvents(after *time.Time) (events []*cloudformation.StackEven
 	if s.StackID == "" {
 		return events, errorNoStackID
 	}
-	err = cfnClient.DescribeStackEventsPages(
-		&cloudformation.DescribeStackEventsInput{
-			StackName: &s.StackID,
-		}, func(page *cloudformation.DescribeStackEventsOutput, lastPage bool) bool {
-			for _, e := range page.StackEvents {
-				if e.Timestamp.UnixNano() > after.UnixNano() {
-					events = append(events, e)
-				}
+	// Find any nested stacks
+	allStacks := []string{s.StackID}
```

Comment:
I feel like adding the stack discovery into here is making `ListEvents` do too much, and makes testing the code more challenging.

Could we break out the nested stack discovery into a separate function and add tests for it, have it called as part of `GetStackInfo` (so it only gets called once per stack) and perform a recursive call to `ListEvents`?

We might need to add a field of `NestedStacks` to the `Stack` struct which should be a slice of `Stack`s. The nested stacks should initialised as the type `Stack`. They would be populated in `GetStackInfo` like so:

```go
s.NestedStacks, err = s.getNestedStacks()
if err != nil {
	return err
}
```

I imagine that the resulting code in `ListEvents` may look something like this:

```go
for _, nestedStack := range s.NestedStacks {
	nestedStackEvents, err := nestedStack.ListEvents(after)
	if err != nil {
		return events, err
	}
	events = append(events, nestedStackEvents...)
}
```

There'd probably need to be a little work around the handling and initialisation of all of this, but I feel this might be a bit more testable, and would also reduce calls to `DescribeStackResources`